### PR TITLE
K8s: Adding ignored pod names as config parameter

### PR DESCRIPTION
**Pull request by eunice98k**
**Description:** <Describe what has changed.>

This PR addresses # 1077 by adding a new parameter to the K8s configuration allowing users to define names of pods to ignore while tagging. Previously, the names are hardcoded and noted as TODO. Modifications were made to move the hardcoded section of the code into `config`.

**Testing:** <Describe what testing was performed and which tests were added.>

- `make` was used to test the new implementation and to ensure no other pre-existing functionality was broken. 

- New unit tests were added to ensure that the default `ignored_pod_names` are `[jaeger-agent, jaeger-collector]` and that users will be able to add additional pod names.

### CODE REVIEW

File: processor/k8sprocessor/config.go [hunk 1]
```diff
@@ -193,3 +197,8 @@ type PodAssociationConfig struct {
 	// e.g. ip, pod_uid, k8s.pod.ip
 	Name string `mapstructure:"name"`
 }
+
+// IgnoredPodNames represent a list of Pod names to ignore
+type PodIgnoredConfig struct {
```
Comment (eddyleelin) [inline processor/k8sprocessor/config.go:202]: The doc name is inconsistent with the variable

File: processor/k8sprocessor/config.go [hunk 2]
```diff
@@ -193,3 +197,8 @@ type PodAssociationConfig struct {
...
+	Name []string `mapstructure:"name"`
```
Comment (eddyleelin) [inline processor/k8sprocessor/config.go:203]: Should this be called Names since it is an array?

File: processor/k8sprocessor/config_test.go [hunk 3]
```diff
@@ -51,6 +51,7 @@ func TestLoadConfig(t *testing.T) {
 		&Config{
 			ProcessorSettings: config.NewProcessorSettings(config.NewID(typeStr)),
 			APIConfig:         k8sconfig.APIConfig{AuthType: k8sconfig.AuthTypeServiceAccount},
+			Ignore:            PodIgnoredConfig{Name: []string{"jaeger-agent", "jaeger-collector"}},
```
Comment (eddyleelin) [inline processor/k8sprocessor/config_test.go:54]: Name -> Names

File: processor/k8sprocessor/config_test.go [hunk 4]
```diff
@@ -105,5 +106,8 @@ func TestLoadConfig(t *testing.T) {
 					Name: "k8s.pod.uid",
 				},
 			},
+			Ignore: PodIgnoredConfig{
+				Name: []string{"jaeger-agent", "jaeger-collector"},
```
Comment (eddyleelin) [inline processor/k8sprocessor/config_test.go:110]: Name -> Names

File: processor/k8sprocessor/factory.go [hunk 5]
```diff
@@ -33,6 +33,7 @@ const (
 
 var kubeClientProvider = kube.ClientProvider(nil)
 var consumerCapabilities = consumer.Capabilities{MutatesData: true}
+var defaultIgnoredNames = PodIgnoredConfig{Name: []string{"jaeger-agent", "jaeger-collector"}}
```
Comment (eddyleelin) [inline processor/k8sprocessor/factory.go:36]: Name -> Names

File: processor/k8sprocessor/kube/client.go [hunk 6]
```diff
@@ -94,6 +96,7 @@ func New(logger *zap.Logger, apiCfg k8sconfig.APIConfig, rules ExtractionRules,
 	}
 
 	c.informer = newInformer(c.kc, c.Filters.Namespace, labelSelector, fieldSelector)
+
```
Comment (eddyleelin) [inline processor/k8sprocessor/kube/client.go:99]: Was there a reason to add this space? 😄

File: processor/k8sprocessor/kube/kube.go [hunk 7]
```diff
@@ -39,10 +39,6 @@ type PodIdentifier string
 
 var (
 	// TODO: move these to config with default values
```
Comment (eddyleelin) [inline processor/k8sprocessor/kube/kube.go:41]: We keep this because the other two vars will need to be updated in the future, too, right?

Comment (eunice98k) [reply processor/k8sprocessor/kube/kube.go:41]: yep!

**Review by alolita — APPROVED** [review]

File: processor/k8sprocessor/options.go [hunk 8]
```diff
@@ -249,3 +249,27 @@ func WithExtractPodAssociations(podAssociations ...PodAssociationConfig) Option
...
+			ignoredNames.Regex = defaultIgnoredNames
+			p.podIgnore = ignoredNames
+			return nil
+		}
+
+		for _, name := range podIgnore.Names {
+			r := regexp.MustCompile(name)
+			ignoredNames.Regex = append(ignoredNames.Regex, r)
+		}
+		p.podIgnore = ignoredNames
```
Comment (Aneurysm9) [inline processor/k8sprocessor/options.go:272]: This allocates a `[]*regexp.Regexp` and the two default regexes, even if they're not going to be used.  With a slight rearrangement that allocation and initialization can be avoided:

```suggestion
		ignoredNames := kube.IgnoredPodNames{}
		names := podIgnore.Names

		if len(names) == 0 {
			names = []*regexp.Regexp{
				regexp.MustCompile(`jaeger-agent`),
				regexp.MustCompile(`jaeger-collector`),
			}
		}

		for _, name := range names {
			ignoredNames.Regex = append(ignoredNames.Regex, regexp.MustCompile(name))
		}
		p.podIgnore = ignoredNames
```

File: processor/k8sprocessor/config.go [hunk 9]
```diff
@@ -193,3 +197,8 @@ type PodAssociationConfig struct {
 	// e.g. ip, pod_uid, k8s.pod.ip
 	Name string `mapstructure:"name"`
 }
+
+// PodIgnoredConfig represent a list of Pod names to ignore
+type PodIgnoredConfig struct {
+	Names []string `mapstructure:"names"`
+}
```
Comment (Aneurysm9) [inline processor/k8sprocessor/config.go:204]:
```suggestion
type PodIgnoredConfig []string
```

The config entry is already called `ignored_pod_names`, so having an additional `names` entry in the map as opposed to just making it a list seems overkill.

File: processor/k8sprocessor/options.go (hunk 8 above)
Comment (eunice98k) [reply processor/k8sprocessor/options.go]: Thanks! Other than names being of type []string, the new commit should reflect your comment. :)

**Pull request closed without merging**